remote/conftest.py

```diff
@@ -0,0 +1,133 @@
+                - auth1
+              hosts:
+                - XX.XX.XXX.XX
+         vcenter:
+          hostname: **Required** -- url to vcenter instance
+          username: **optional** -- will first try environment variable $VCUSER
+          password: **optional** -- will first try environment variable $VCPASS
+
+    *Warning*
+    It is a bad idea to have multiple sessions of pytest running
+    concurrently against the same hosts/config file.
+    If one session ends before another, it will powerdown the
+    machines leaving the other one to fail.
+    """
+    from pyVim.connect import SmartConnect, Disconnect
+    import ssl
+
+    cfg = config.get_config()
+    s = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+    s.verify_mode = ssl.CERT_NONE
+    # try and make a secure connection first
+    if 'vcenter' in cfg.keys():
+        if os.getenv('VCUSER', False):
+            vcuser = os.environ['VCUSER']
+        else:
+            vcuser = cfg['vcenter']['username']
+        if os.getenv('VCPASS', False):
+            vcpassword = os.environ['VCPASS']
+        else:
+            vcpassword = cfg['vcenter']['password']
+        try:
+            c = SmartConnect(
+                host=cfg['vcenter']['hostname'],
+                user=vcuser,
+                pwd=vcpassword
+            )
+        except ssl.SSLError:
+            c = SmartConnect(
+                host=cfg['vcenter']['hostname'],
+                user=vcuser,
+                pwd=vcpassword,
+                sslContext=s
+            )
+        # these index choices are particular to our VCenter setup
+        host_folder = c.content.rootFolder.childEntity[DATA_CENTER].hostFolder
+        vm_host = host_folder.childEntity[CLUSTER].host[HOST]
+        # the host has a property "vm" which is a list of VMs
+        vms = vm_host.vm
+        for host in cfg['rho']['hosts']:
+            if ('provider' in host.keys()) and (
+                    host['provider'] == 'vcenter'):
+                for vm in vms:
+                    if vm.name == host['hostname']:
+                        if vm.runtime.powerState == 'poweredOff':
+                            vm.PowerOnVM_Task()
+        # need to wait for machines to boot and start sshd
+        sleep(30)
+
+    def shutdown_systems():
+        """Turn off hosts at the end of a session.
+
+        This runs once at the end of each session.
+        """
+        cfg = config.get_config()
+        for host in cfg['rho']['hosts']:
+            if ('provider' in host.keys()) and (host['provider'] == 'vcenter'):
+                for vm in vms:
+                    if vm.name == host['hostname']:
+                        if vm.runtime.powerState == 'poweredOn':
+                            vm.PowerOffVM_Task()
+        Disconnect(c)
+
+        #  this line makes this part of the fixture only run at end of session
+        #  pytest provides us this request object
+        request.addfinalizer(shutdown_systems)  # noqa: F821
```

Comment:
I am curious how this request is passed. It seems that it is not being called as it should be one level back in the indentation to be on the manage_systems body. Or am I missing something?

Reply:
Oops! That got lost in the shuffle. Fixing it now.